pkg/yurthub/kubernetes/serializer/serializer.go

```diff
@@ -40,25 +45,145 @@ type SerializerManager struct {
+func NewUnstructuredNegotiatedSerializer() runtime.NegotiatedSerializer {
+	return UnstructuredNegotiatedSerializer{
+		scheme:  scheme.Scheme,
+		typer:   unstructuredscheme.NewUnstructuredObjectTyper(),
+		creator: NewUnstructuredCreator(),
+	}
+}
+
+func (s UnstructuredNegotiatedSerializer) SupportedMediaTypes() []runtime.SerializerInfo {
+	return []runtime.SerializerInfo{
+		{
+			MediaType:        "application/json",
+			MediaTypeType:    "application",
+			MediaTypeSubType: "json",
+			EncodesAsText:    true,
+			Serializer:       json.NewSerializerWithOptions(json.DefaultMetaFactory, s.creator, s.typer, json.SerializerOptions{}),
+			PrettySerializer: json.NewSerializerWithOptions(json.DefaultMetaFactory, s.creator, s.typer, json.SerializerOptions{Pretty: true}),
+			StreamSerializer: &runtime.StreamSerializerInfo{
+				EncodesAsText: true,
+				Serializer:    json.NewSerializerWithOptions(json.DefaultMetaFactory, s.creator, s.typer, json.SerializerOptions{}),
+				Framer:        json.Framer,
+			},
+		},
+		{
+			MediaType:        "application/yaml",
+			MediaTypeType:    "application",
+			MediaTypeSubType: "yaml",
+			EncodesAsText:    true,
+			Serializer:       json.NewSerializerWithOptions(json.DefaultMetaFactory, s.creator, s.typer, json.SerializerOptions{Yaml: true}),
+		},
+	}
+}
+
+//EncoderForVersion do nothing, but returns a encoder,
+//if the object is unstructured, the encoder will encode object without conversion
+func (s UnstructuredNegotiatedSerializer) EncoderForVersion(encoder runtime.Encoder, gv runtime.GroupVersioner) runtime.Encoder {
+	return versioning.NewDefaultingCodecForScheme(s.scheme, encoder, nil, gv, nil)
+
+}
+
+//DecoderToVersion do nothing, and returns a decoder that does not do conversion
+func (s UnstructuredNegotiatedSerializer) DecoderToVersion(decoder runtime.Decoder, _ runtime.GroupVersioner) runtime.Decoder {
+	return WithVersionDecoder{
+		Decoder: decoder,
+	}
+}
+
+type unstructuredCreator struct{}
+
+// NewUnstructuredCreator returns a simple object creator that always returns an Unstructured or UnstructuredList
+func NewUnstructuredCreator() runtime.ObjectCreater {
+	return unstructuredCreator{}
+}
+
+func (c unstructuredCreator) New(kind schema.GroupVersionKind) (runtime.Object, error) {
+	if strings.HasSuffix(kind.Kind, "List") {
+		ret := &unstructured.UnstructuredList{}
+		ret.SetGroupVersionKind(kind)
+		return ret, nil
+	} else {
+		ret := &unstructured.Unstructured{}
+		ret.SetGroupVersionKind(kind)
+		return ret, nil
 	}
 }
 
 // CreateSerializers create a *rest.Serializers for encoding or decoding runtime object
-func (sm *SerializerManager) CreateSerializers(contentType, group, version string) (*rest.Serializers, error) {
-	mediaTypes := sm.NegotiatedSerializer.SupportedMediaTypes()
+func (sm *SerializerManager) CreateSerializers(contentType, group, version, apiPrefix string) (*rest.Serializers, error) {
+	var mediaTypes []runtime.SerializerInfo
+	if apiPrefix == "api" {
```

Comment:
I think there are some normal gvk with apiPrefix="api" and contentType="xxx.protobuf.xxx", and `CreateSerializers` can not handle them.

Reply:
my suggestion: use gvr + restMapper to get gvk, so select serializer based on gvk.

Reply:
code is already updated, thanks for review